File: src/0394_decode_string.js
function decodeString(s) {
    const stack = [];
    for (const char of s) {
        if (char !== ']') {
            stack.push(char);
        } else {
            let subStr = '';
            while (stack[stack.length - 1] !== '[') subStr = stack.pop() + subStr;
            stack.pop();
            let multiplier = '';
            while (stack.length && /\d/.test(stack[stack.length - 1])) multiplier = stack.pop() + multiplier;
            stack.push(subStr.repeat(Number(multiplier)));
        }
    }
    return stack.join('');
}

module.exports = { decodeString };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(decodeString("3[a]2[bc]"), "aaabcbc");
    assert.deepStrictEqual(decodeString("3[a2[c]]"), "accaccacc");
}

File: src/0394_decode_string.rs
struct Solution;

impl Solution {
    pub fn decode_string(s: String) -> String {
        let mut stack = vec![];

        for ch in s.chars() {
            if ch != ']' {
                stack.push(ch.to_string());
            } else {
                let mut substr = String::new();
                while stack.last() != Some(&"[".to_string()) {
                    substr = stack.pop().unwrap().to_string() + &substr;
                }
                stack.pop();

                let mut multiplier = String::new();
                while !stack.is_empty() && stack.last().unwrap().parse::<i32>().is_ok() {
                    multiplier = stack.pop().unwrap().to_string() + &multiplier;
                }

                stack.push(substr.repeat(multiplier.parse::<usize>().unwrap()));
            }
        }
        stack.join("")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn main_asserts() {
        assert_eq!(
            Solution::decode_string("3[a]2[bc]".to_string()),
            "aaabcbc".to_string()
        );
        assert_eq!(
            Solution::decode_string("3[a2[c]]".to_string()),
            "accaccacc".to_string()
        );
    }
}
